system/acpi.c

```diff
@@ -0,0 +1,345 @@
+            uintptr_t addr = *tab_ptr++;  // read the next table entry
+            uint32_t *ptr = (uint32_t *)map_region(addr, sizeof(uint32_t), true);
+
+            if (ptr && *ptr == table_signature) {
+                return addr;
+            }
+        }
+    } else {
+        rt = (rsdt_header_t *)map_region(rp->rsdt_addr, sizeof(rsdt_header_t), true);
+        if (rt == NULL) {
+            return 0;
+        }
+        // Validate the RSDT.
+        if (*(uint32_t *)rt != RSDTSignature) {
+            return 0;
+        }
+        rt = (rsdt_header_t *)map_region(rp->rsdt_addr, rt->length, true);
+        if (rt == NULL || acpi_checksum(rt, rt->length) != 0) {
+            return 0;
+        }
+        // Scan the RSDT for a pointer to the table we're looking for.
+        uint32_t *tab_ptr = (uint32_t *)((uint8_t *)rt + sizeof(rsdt_header_t));
+        uint32_t *tab_end = (uint32_t *)((uint8_t *)rt + rt->length);
+
+        while (tab_ptr < tab_end) {
+            uintptr_t addr = *tab_ptr++;  // read the next table entry
+            uint32_t *ptr = (uint32_t *)map_region(addr, sizeof(uint32_t), true);
+
+            if (ptr && *ptr == table_signature) {
+                return addr;
+            }
+        }
+    }
+
+    return 0;
+}
+
+static bool parse_fadt(uintptr_t fadt_addr)
+{
+    // FADT is a very big & complex table and we only need a few data.
+    // We use byte offset instead of a complete struct.
+
+    // FADT Header is identical to RSDP Header
+    rsdt_header_t *fadt = (rsdt_header_t *)fadt_addr;
+
+    // Validate FADT
+    if (fadt == NULL || acpi_checksum(fadt, fadt->length) != 0) {
+        return false;
+    }
+
+    // Get ACPI Version
+    acpi_config.ver_maj = fadt->revision;
+
+    if (fadt->length > FADT_MINOR_REV_OFFSET) {
+        acpi_config.ver_min = *(uint8_t *)(fadt_addr+FADT_MINOR_REV_OFFSET) & 0xF;
+    }
+
+    // Get Old PM Base Address (32bit IO)
+    acpi_config.pm_addr  = *(uint32_t *)(fadt_addr+FADT_PM_TMR_BLK_OFFSET);
+    acpi_config.pm_is_io = true;
+
+#ifdef __x86_64__
+    acpi_gen_addr_struct *rt;
+
+    // Get APIC Timer Address
+    if (fadt->length > FADT_X_PM_TMR_BLK_OFFSET) {
+        rt = (acpi_gen_addr_struct *)map_region(fadt_addr+FADT_X_PM_TMR_BLK_OFFSET, sizeof(acpi_gen_addr_struct), true);
+
+        acpi_config.pm_is_io = (rt->address_space == 1) ? true : false;
+
+        if (rt->address != 0) {
+            acpi_config.pm_addr = rt->address;
+        }
+    }
+#endif
+
+    return true;
+}
+
+//------------------------------------------------------------------------------
+// Public Functions
+//------------------------------------------------------------------------------
+
+int acpi_checksum(const void *data, int length)
+{
+    uint8_t sum = 0;
+
+    uint8_t *ptr = (uint8_t *)data;
+    while (length--) {
+        sum += *ptr++;
+    }
+    return sum;
+}
+
+void acpi_init(void)
+{
+    // Find ACPI RDST Table Address
+    acpi_config.rsdp_addr = find_rsdp();
+
```

Comment:
Should test for failure (zero address) here, otherwise the following calls to find_acpi_table will dereference an invalid pointer.